ddtrace/internal/service.py

```diff
@@ -56,5 +76,87 @@ def stop(self):
         self.status = ServiceStatus.STOPPED
 
     @staticmethod
-    def join(timeout=None):
+    def join(
+        timeout=None,  # type: typing.Optional[float]
+    ):
+        # type: (...) -> None
         """Join the service once stopped."""
+
+    def copy(self):
+        # type: () -> Service
+        return attr.evolve(self)
+
+
+class ForksafeService(object):
+    """A manager that can handle the lifecycle of a service in a process.
+
+    This allows to handle a service that should keep running when, e.g., the Python process forks.
+    """
+
+    SERVICE_CLASS = None  # type: typing.ClassVar[typing.Type[Service]]
+
+    def __init__(
+        self,
+        *args,  # type: typing.Any
+        **kwargs  # type: typing.Any
+    ):
+        # type: (...) -> None
+        super(ForksafeService, self).__init__()
+        self._service = self.SERVICE_CLASS(*args, **kwargs)  # type: ignore[call-arg]
+
+    def start(self, *args, **kwargs):
+        # type: (...) -> None
+        """Start the service.
+
+        :param stop_on_exit: Whether to stop the service and wait for shutdown on exit.
+        :param start_in_children: Whether to start the service in child processes.
+        """
+        stop_on_exit = kwargs.pop("stop_on_exit", True)
+        start_in_children = kwargs.pop("start_in_children", True)
+
+        if start_in_children:
+            try:
+                uwsgi.check_uwsgi(self._restart_on_fork, atexit=self.stop if stop_on_exit else None)
+            except uwsgi.uWSGIMasterProcess:
+                # Do nothing, the start() method will be called in each worker subprocess
+                return
+
+        self._service.start(*args, **kwargs)
+
+        if stop_on_exit:
+            atexit.register(self.stop)
+
+        if start_in_children:
+            forksafe.register(self._restart_on_fork)
+
+    def stop(self):
+        # type: () -> None
+        """Stop the service."""
+        # Check for the service status: this *should* not be necessary per-say, but in the case `atexit.unregister` is
+        # unavailable, we're sure not to call stop twice.
+        if self._service.status == ServiceStatus.RUNNING:
```

Comment:
Is there a possible race condition here? We handle starting race conditions but seems like it'd also make sense to do it on stop() too (unless we state that stop() may be called more than once)

Reply:
We don't handle race conditions in `start` and this whole class is **not** thread-safe.

Does it have to be thread-safe?

In the case of the `Profiler`, I'm OK stating that the `Profiler` object is not thread-safe and that you cannot call `start()` multiple times from different threads.

Reply:
Oh yeah I meant we handle the underlying start race condition in `Service` but we don't with stop.

Yeah I don't think it has to be thread-safe but wanted to make sure that we were aware of it - maybe worth indicating it explicitly in the docstring?

> In the case of the Profiler, I'm OK stating that the Profiler object is not thread-safe and that you cannot call start() multiple times from different threads.

:+1: agreed

Reply:
Sure, I've added something in the doc string.